ADAL/src/cache/ios/ADKeychainTokenCache.m

```diff
@@ -435,26 +435,18 @@ - (BOOL)removeAllForUserId:(NSString * __nonnull)userId
     return YES;
 }
 
-- (BOOL)removeAllForUserId:(NSString *)userId error:(ADAuthenticationError *__autoreleasing  _Nullable *)error
+- (BOOL)wipeAllItemsForUserId:(NSString *)userId error:(ADAuthenticationError *__autoreleasing  _Nullable *)error
 {
     AD_LOG_WARN(nil, @"Removing all items for user.");
     AD_LOG_WARN_PII(nil, @"userId <%@>", userId);
 
-    NSArray *items = [self allItems:nil];
-    if (!items)
-    {
-        return NO;
-    }
-    
-    for (ADTokenCacheItem *item in items)
+    @synchronized(self)
     {
-        if ([userId isEqualToString:[[item userInformation] userId]]
-            && ![self removeItem:item error:error])
-        {
-            return NO;
-        }
+        NSMutableDictionary *query = [self queryDictionaryForKey:nil userId:userId additional:nil];
```

Comment:
Avoid this as it has different side effects then you are taking into account here, specifically if you look at this part:

```objective-c
    NSMutableDictionary* defaultQuery =
    [@{
       (id)kSecClass : (id)kSecClassGenericPassword,
       (id)kSecAttrGeneric : [s_libraryString dataUsingEncoding:NSUTF8StringEncoding]
       } mutableCopy];
    
    // Use a different generic attribute so that past versions of ADAL don't trip up on this entry
    NSMutableDictionary* defaultTombstoneQuery =
    [@{
       (id)kSecClass : (id)kSecClassGenericPassword,
       (id)kSecAttrGeneric : [s_tombstoneLibraryString dataUsingEncoding:NSUTF8StringEncoding]
       } mutableCopy];
```

This query is only going to delete non-tombstones. We want it to delete both. It'd be better and clearer to simply create the query dictionary here with the appropriate attributes:

```objective-c
NSDictionary *query =
@{ (id)kSecClass : (id)kSecClassGenericPassword,
      (id)kSecAccount: [userId adBase64UrlEncode],
      (id)kSecAttrAccessGroup: _sharedGroup };
```